rfd/0079-oidc-joining.md

```diff
@@ -0,0 +1,210 @@
+
+## Why
+
+This feature reduces the friction involved in dynamically adding many new nodes to Teleport on a variety of platforms. This is also more secure, as the user does not need to distribute a token which is liable to exfilitration.
+
+Whilst multiple providers offer OIDC identities to workloads running on their platform, we will start by targetting GCP since this represents a key platform for Teleport, and is also well documented and easy to test on. However, the work towards this feature will also enable us to simply add other providers that support OIDC workload identity (see the references for more), this particularly ties into Machine ID goals as we aim to support several CI/CD providers that offer workload identity.
+
+## Details
+
+The work on OIDC joining is broken down into two parts:
+
+- Support in the Auth server for trusting an OP issued JWT. This work will have a general base that can be shared between all providers, and then have a small, provider-specific part to provide additional validation that lets us guide users to correct and safe configurations.
+- Support in nodes for fetching their workload identity token from their environment. This work will be specific to each platform we intend to support.
+
+OIDC supports multiple types of token (`id_token`: a JWT encoding information about the identity, which can be verified using the issuer's public key and `access_token`: an opaque token that can be used with a `userinfo` endpoint on the issuer to obtain information about the identity). However, in the case of workload identities, `id_token` is the most prevelant. For this reason, our initial implementation will solely support `id_token`.
+
+### Auth server support
+
+We will re-use the existing endpoints around joining as much as possible. This means that the main entry-point for joining will be the existing `RegisterUsingToken` method.
+
+We will introduce a new token type for each provider. In the case of GCP this will be `oidc-gcp`.
+
+Registration flow:
+
+1. Client is configured by the user to use `oidc-gcp` joining. The client will then query the Google Metadata Service to obtain a JWT.
+2. The client will call the `RegisterUsingToken` endpoint, providing the OIDC JWT that it has collected, and specifying the name of the Teleport provisioning token which should be used to verify it.
+3. The server will attempt to fetch the Token resource for the specified token.
+4. The server will check JWT header to ensure the `alg` is one we have allow-listed (RS[256, 384, 512])
+5. The server will check the `kid` of the JWT header, and obtain the relevant JWK from the cache or from the specified issuers well-known JWKS endpoint. It will then use the JWK to validate the token has been signed by the issuer.
+6. Other key claims of the JWT will be validated:
+  - Ensure the audience is correct (for providers where the audience can be customised, this should be the Teleport cluster name.)
```

Comment:
Rather than special-casing the `audience` validation, could we include it as a claim in the `allow` rules? I could see this being more flexible if on a certain provider the audience can't be controlled, the Teleport admin can decide whether or not to validate the audience and which value to look for.

Reply:
Now that I've redesigned the specification to have provider specific token types (e.g `oidc-gcp` rather than a generic `oidc-jwt` type), I feel that making the admin configure what this value should be is risky and burdens them with unnecessary configuration. We will know whether or not the provider we are supporting has the ability to customise this audience, and in that case, we will want to force it to be set to the name of the cluster. In the case it can't be customised, we can check that this value is the expected value for that provider.

Reply:
Good point. I was thinking there could be some provider where the audience is a value we can't control, but of course we can add the config option for that specific provider type if it was ever necessary